api/grpcserver/globalstate_service.go

```diff
@@ -270,33 +270,53 @@ func (s GlobalStateService) AccountDataStream(in *pb.AccountDataStreamRequest, s
 	filterReceipt := in.Filter.AccountDataFlags&uint32(pb.AccountDataFlag_ACCOUNT_DATA_FLAG_TRANSACTION_RECEIPT) != 0
 
 	// Subscribe to the various streams
-	var accountCh, rewardsCh, receiptsCh <-chan interface{}
-
+	var (
+		accountCh       <-chan interface{}
+		rewardsCh       <-chan interface{}
+		receiptsCh      <-chan interface{}
+		accountBufFull  <-chan struct{}
+		rewardsBufFull  <-chan struct{}
+		receiptsBufFull <-chan struct{}
+	)
 	if filterAccount {
 		if accountSubscription := events.SubscribeAccount(); accountSubscription != nil {
 			defer closeSubscription(accountSubscription)
-			accountCh = consumeEvents(stream.Context(), accountSubscription.Out())
+
+			accountCh, accountBufFull = consumeEvents(stream.Context(), accountSubscription.Out())
 		}
 	}
 	if filterReward {
 		if rewardsSubscription := events.SubscribeRewards(); rewardsSubscription != nil {
 			defer closeSubscription(rewardsSubscription)
-			rewardsCh = consumeEvents(stream.Context(), rewardsSubscription.Out())
+			rewardsCh, rewardsBufFull = consumeEvents(stream.Context(), rewardsSubscription.Out())
 		}
 	}
 	if filterReceipt {
 		if receiptsSubscription := events.SubscribeReceipts(); receiptsSubscription != nil {
 			defer closeSubscription(receiptsSubscription)
-			receiptsCh = consumeEvents(stream.Context(), receiptsSubscription.Out())
+			receiptsCh, receiptsBufFull = consumeEvents(stream.Context(), receiptsSubscription.Out())
 		}
 	}
 
 	for {
 		select {
+		case <-accountBufFull:
+			log.Info("account buffer is full, shutting down")
+			return status.Error(codes.Canceled, errAccountBufferFull)
+		case <-rewardsBufFull:
+			log.Info("rewards buffer is full, shutting down")
+			return status.Error(codes.Canceled, errRewardsBufferFull)
+		case <-receiptsBufFull:
+			log.Info("receipts buffer is full, shutting down")
+			return status.Error(codes.Canceled, errReceiptsBufferFull)
```

Comment:
if all the result is the same when ANY buffer is full, can we simply pass in one channel for any of the go routine to close?
that way, other goroutines can check if that channel is closed and exit?

Reply:
We could, but in this case, logs won't show what buffer became full, which can make debugging harder

Reply:
the one that closes the channel when its buffer is full can log it. i dont see how that's an issue.